mainnet-contracts/src/PufferProtocol.sol

```diff
@@ -284,10 +276,87 @@ contract PufferProtocol is IPufferProtocol, AccessManagedUpgradeable, UUPSUpgrad
         --$.nodeOperatorInfo[node].pendingValidatorCount;
         ++$.nodeOperatorInfo[node].activeValidatorCount;
 
+        // Update numBatches now that validator becomes active
+        $.nodeOperatorInfo[node].numBatches += $.validators[moduleName][index].numBatches;
+
         // Mark the validator as active
         $.validators[moduleName][index].status = Status.ACTIVE;
     }
 
+    /**
+     * @inheritdoc IPufferProtocol
+     * @dev Restricted to Node Operators
+     */
+    function requestConsolidation(bytes32 moduleName, uint256[] calldata srcIndices, uint256[] calldata targetIndices)
+        external
+        payable
+        virtual
+        restricted
+    {
+        if (srcIndices.length == 0) {
+            revert InputArrayLengthZero();
+        }
+        if (srcIndices.length != targetIndices.length) {
+            revert InputArrayLengthMismatch();
+        }
+
+        ProtocolStorage storage $ = _getPufferProtocolStorage();
+
+        bytes[] memory srcPubkeys = new bytes[](srcIndices.length);
+        bytes[] memory targetPubkeys = new bytes[](targetIndices.length);
+        Validator storage validatorSrc;
+        Validator storage validatorTarget;
+        for (uint256 i = 0; i < srcPubkeys.length; i++) {
+            require(srcIndices[i] != targetIndices[i], InvalidValidator());
+            validatorSrc = $.validators[moduleName][srcIndices[i]];
+            require(validatorSrc.node == msg.sender && validatorSrc.status == Status.ACTIVE, InvalidValidator());
+            srcPubkeys[i] = validatorSrc.pubKey;
+            validatorTarget = $.validators[moduleName][targetIndices[i]];
+            require(validatorTarget.node == msg.sender && validatorTarget.status == Status.ACTIVE, InvalidValidator());
+            targetPubkeys[i] = validatorTarget.pubKey;
+
+            // Update accounting
+            validatorTarget.bond += validatorSrc.bond;
+            validatorTarget.numBatches += validatorSrc.numBatches;
+
+            _deleteValidator(validatorSrc);
+            // Node info needs no update since all stays in the same node operator
+        }
+
+        $.modules[moduleName].requestConsolidation{ value: msg.value }(srcPubkeys, targetPubkeys);
+
+        emit ConsolidationRequested(moduleName, srcPubkeys, targetPubkeys);
+    }
+
+    /**
+     * @inheritdoc IPufferProtocol
+     * @dev Restricted to Node Operators
+     */
+    function requestWithdrawal(bytes32 moduleName, uint256[] calldata indices, uint64[] calldata gweiAmounts)
+        external
+        payable
+        restricted
+    {
+        if (indices.length == 0) {
+            revert InputArrayLengthZero();
+        }
+        if (indices.length != gweiAmounts.length) {
+            revert InputArrayLengthMismatch();
+        }
```

Comment:
Related to the PufferModuleManager validations in `requestWithdrawal` comment. If we remove those, we will save some money on Gas, but if we get a bad input in the function, it will either panic / revert somewhere. It might not be the best UX for the reverting txs, but do we care about that as much? Since it is expected that 99% txs will be good, it will waste some money for those 99% to have double validations.